perf_events.go

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2019 Dropbox, Inc.
+// Full license can be found in the LICENSE file.
+
+package goebpf
+
+/*
+#ifdef __linux__
+#include <linux/perf_event.h>
+#define PERF_EVENT_HEADER_SIZE		(sizeof(struct perf_event_header))
+#else
+// mocks for Mac
+#define PERF_EVENT_HEADER_SIZE		8
+#define PERF_RECORD_SAMPLE			9
+#define PERF_RECORD_LOST			2
+#endif
+
+*/
+import "C"
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sync"
+)
+
+// PerfEvent is a way to interact with Linux's PerfEvents for eBPF case.
+type PerfEvents struct {
+	// Statistics
+	EventsReceived   int
+	EventsLost       int
+	EventsUnknowType int
+
+	// PollTimeoutMs is timeout for blocking call of poll()
+	// Defaults to 100ms
+	PollTimeoutMs int
+
+	perfMap        Map
+	updatesChannel chan []byte
+	stopChannel    chan bool
+	wg             sync.WaitGroup
+
+	handlers []*perfEventHandler
+}
+
+// Go definition for C structs from
+// http://man7.org/linux/man-pages/man2/perf_event_open.2.html
+// struct perf_event_header {
+//     __u32   type;
+//     __u16   misc;
+//     __u16   size;
+// }
+type perfEventHeader struct {
+	Type uint32
+	Misc uint16
+	Size uint16
+}
+
+// struct perf_event_lost {
+//     uint64_t id;
+//     uint64_t lost;
+// not added: struct sample_id sample_id;
+// }
+type perfEventLost struct {
+	Id   uint64
+	Lost uint64
+}
+
+// NewPerfEvents creates new instance of PerfEvents for eBPF map "m".
+// "m" must be a type of "MapTypePerfEventArray"
+func NewPerfEvents(m Map) (*PerfEvents, error) {
+	if m.GetType() != MapTypePerfEventArray {
+		return nil, fmt.Errorf("Invalid map type '%v'", m.GetType())
+	}
+
+	return &PerfEvents{
+		perfMap:       m,
+		PollTimeoutMs: 100,
+	}, nil
+}
+
+// StartForAllProcessesAndCPUs starts PerfEvent polling on all CPUs for all system processes
+// This mode requires specially organized map: index matches CPU ID.
+// "bufferSize" is ring buffer size for perfEvents. Per CPU.
+// All updates will be sent into returned channel.
+func (pe *PerfEvents) StartForAllProcessesAndCPUs(bufferSize int) (<-chan []byte, error) {
+	// Get CPU count
+	nCpus, err := GetNumOfPossibleCpus()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create perfEvent handler for all possible CPUs
+	pe.handlers = nil
```

Comment:
nit: `pe.handlers = make([]*.., nCpus)` to avoid allocations in loop.

Reply:
This is one time call function so I'd consider this is as minor.
Anyway refactored to make it more readable.